File: C#/79_WordSearch.cs
public class Solution {
    // 10:52
    public bool Exist(char[][] board, string word) {
        for (int i = 0; i < board.Length; i++){
            for (int j = 0; j < board[i].Length; j++){
                if (board[i][j] != word[0]) continue;
                if (Check(board, word, 0, (i,j), new HashSet<(int,int)>()))
                    return true;
            }
        }
        return false;
    }
    
    private bool Check(char[][] board, string w, int index, (int,int) start, HashSet<(int,int)> used) {
        used.Add(start);
        (int x, int y) = start;
        if (index == w.Length - 1)
            return true;
        
        bool foundPath = false;
        
        // Check top..
        if (x > 0 && w[index+1] == board[x-1][y] && !used.Contains((x-1,y))) {
            foundPath = Check(board, w,  index + 1, (x-1,y), used);
            if (foundPath) return true;
            used.Remove((x-1,y));
        }
        // Check bottom..
        if (x < board.Length-1 && w[index+1] == board[x+1][y] && !used.Contains((x+1,y))) {
            foundPath = Check(board, w, index + 1, (x+1,y), used);
            if (foundPath) return true;
            used.Remove((x+1,y));
        }
            
        // Check left..
        if (y > 0 && w[index+1] == board[x][y-1] && !used.Contains((x,y-1))) {
            foundPath = Check(board, w, index + 1, (x,y-1), used);
            if (foundPath) return true;
            used.Remove((x,y-1));
        }
        
        // Check right..
        if (y < board[x].Length-1 && w[index+1] == board[x][y+1] && !used.Contains((x,y+1))) {
            foundPath = Check(board, w, index + 1, (x,y+1), used);
            if (foundPath) return true;
            used.Remove((x,y+1));
        }
        return false;
    }
}

File: Python/79_WordSearch.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        used = set()
        
        def DFS(row, col, index):
            if index == len(word): return True
            
            offsets = [(0,1), (0,-1), (1,0), (-1,0)]
            used.add((row, col))
            
            for o_x, o_y in offsets:
                x = o_x + row
                y = o_y + col
                if (x,y) in used: continue
                if 0 <= x < len(board) and 0 <= y < len(board[x]) and board[x][y] == word[index]:
                    if DFS(x, y, index + 1): return True
            
            used.remove((row, col))
            return False
        
        
        for row in range(len(board)):
            for col in range(len(board[row])):
                if board[row][col] == word[0] and DFS(row, col, 1):
                    return True
        
        return False
